test/topology_raft_disabled/test_raft_upgrade.py

```diff
@@ -8,91 +8,345 @@
-            if rs:
-                value = rs[0].value
-                if value == 'use_post_raft_procedures':
-                    break
-                else:
-                    logging.info(f"Upgrade not finished yet on server {host}, state: {value}")
-            else:
-                logging.info(f"Upgrade not finished yet on server {host}, no upgrade state written")
-            await asyncio.sleep(1)
+    logging.info(f"Driver reconnected, hosts: {hosts}. Waiting until upgrade finishes")
+    await asyncio.gather(*(wait_until_upgrade_finishes(cql, h, time.time() + 60) for h in hosts))
 
     logging.info("Upgrade finished. Creating a new table")
-
     table = await random_tables.add_table(ncolumns=5)
-    table_name = table.full_name
 
     logging.info("Checking group0_history")
     rs = await cql.run_async("select * from system.group0_history")
     assert(rs)
     logging.info(f"group0_history entry description: '{rs[0].description}'")
-    assert(table_name in rs[0].description)
+    assert(table.full_name in rs[0].description)
+
+
+@pytest.mark.asyncio
+@log_run_time
+async def test_raft_upgrade_with_node_remove(manager: ManagerClient, random_tables: RandomTables):
+    servers = await manager.running_servers()
+    srv1, *others = servers
+
+    srv1_gen = await manager.api.get_gossip_generation_number(others[0], srv1)
+    logging.info(f"Gossip generation number of {srv1} seen from {others[0]}: {srv1_gen}")
+
+    logging.info(f"Enabling Raft on {srv1} and restarting")
+    await enable_raft_and_restart(manager, srv1)
+
+    # Before continuing, ensure that another node has gossiped with srv1
+    # after srv1 has restarted. Then we know that the other node learned about srv1's
+    # supported features, including SUPPORTS_RAFT.
+    logging.info(f"Waiting until {others[0]} gossips with {srv1}")
+    await wait_for_gossip_gen_increase(manager.api, srv1_gen, others[0], srv1, time.time() + 60)
+
+    srv1_host_id = await manager.get_host_id(srv1)
+    logging.info(f"Obtained host ID of {srv1}: {srv1_host_id}")
+
+    logging.info(f"Stopping {srv1}")
+    await manager.server_stop_gracefully(srv1)
+
+    logging.info(f"Enabling Raft on {others} and restarting")
+    await asyncio.gather(*(enable_raft_and_restart(manager, srv) for srv in others))
+    cql = await reconnect_driver(manager)
+
+    logging.info(f"Cluster restarted, waiting until driver reconnects to every server except {srv1}")
+    hosts = await wait_for_cql_and_get_hosts(cql, others, time.time() + 60)
+    logging.info(f"Driver reconnected, hosts: {hosts}")
+
+    logging.info(f"Removing {srv1} using {others[0]}")
+    await manager.remove_node(others[0], srv1, srv1_host_id)
+
+    logging.info("Waiting until upgrade finishes")
+    await asyncio.gather(*(wait_until_upgrade_finishes(cql, h, time.time() + 60) for h in hosts))
+
+
+@pytest.mark.asyncio
```

Comment:
Please write the test plan here, i.e. what you're doing in the test step by step and what it is supposed to test. The same is true for other tests in this file. The idea of the test should be given for any non-trivial test.
Thanks.

Reply:
Done